### ISSUE TITLE
Implement event logging in Sourcegraph for admins

### ISSUE THREAD

**Issue by dadlerj** [sourcegraph/sourcegraph#5486]
This project is copied from https://github.com/sourcegraph/sourcegraph/issues/3583#issuecomment-521432491

See the bottom for an update on concrete next steps to consider this project complete.

---

Thanks for all of the feedback, everyone!

This is not going to be as pretty as the project scoping docs that Product usually puts together, so please let me know where more info would be helpful!

**Background/problems**

From my point of view, this project is about building a common backend that supports multiple frontend products/use cases, not all of which are needed today.

The problems:

* The current site usage statistics page is built on an extremely brittle and un-extensible redis-based backend. Adding a new metric almost always requires modifying/migrating the redis data structure (and all of the stored data), or adding an entirely new one.

* As a result, capturing any new metrics, or adding any features based on user action aggregation (such as https://github.com/sourcegraph/sourcegraph/issues/3333 which I added a couple iterations ago, https://github.com/sourcegraph/sourcegraph/issues/2069 which Beyang added a few iterations ago, @vanesa's work on https://github.com/sourcegraph/sourcegraph/issues/2348, soon https://github.com/sourcegraph/sourcegraph/issues/5088, and any future work on audit logs) requires:
  1. Adding a new storage backend, new backend event handlers, and new data aggregators (see the churn in the `usagestats` package).
  2. Deciding on a data structure, writing new tests, testing for performance at massive scale.
  3. Adding the new frontend code to custom handle the new event(s).

**Definition of success/Scope**

The concrete outcome I'd like to reach is:

- "All" frontend user actions are logged inside of the Sourcegraph instance. This includes everything from "user viewed a page", to "user clicked a button", to "user hovered over a symbol" — along the lines of a Google Analytics, or what we use BigQuery/Looker for for Sourcegraph.com metrics. These user actions (or "events") that are logged can accept an arbitrary text (or json?) argument.
- "All" is quoted above to indicate that it's not really all events :) — rather, only the things we care about would actually be logged. See [all calls to `eventLogger.log` ](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+eventlogger.log%28) in the frontend code to get a sense for the scale here
- All of these events would pass through a common GraphQL endpoint and event handler on the backend that would add them into the new database.
- All of these events are stored in a backend database containing:
  - `name`: the name of the event
  - `argument` (or whatever): the string argument
  - `url`: the URL on the page when the event was logged
  - `userID`: the user that did the action
  - `anonymousUserID`: a UUID stored in a cookie that we use as a way to aggregate actions by anonymous users (when the instance has `"auth.public": true` set). This is already generated and used in the existing usagestats package.
   - `timestamp`: the timestamp of the action
- From this raw database we could easily aggregate data as needed to generate each of the features/products described above, including:
  - Site usage statistics — e.g. a count of monthly active users would be as simple as: `SELECT COUNT(DISTINCT user_id) FROM events WHERE timestamp > $start_of_month AND timestamp < $end_of_month`
  - Extension usage (@vanesa's project) — a count of unique users by extension would be as simple as `SELECT user_id, argument FROM events WHERE name='ExtensionActivated'` to get the list, then loop over each record to build the list.
  - Audit logs — e.g., a list of actions that match some list of "audit loggable" events, such as `SELECT name, user_id, timestamp FROM events WHERE name IN ('SettingsChanged', 'ConfigChanged', 'RepoViewed',  ...)`

**Out of scope**

Note that actually creating the features/products listed aboved is NOT part of the definition of success — rather, once the data store is available we would begin to add/migrate these. 

Any thoughts?

---

Now that @unknwon's perf testing results are complete, my take is that the following is required from here:

- [ ] Actually implement the db table in Postgres (as well as any supporting service infrastructure, such as the gc to dump data after 93 days, etc.)
- [ ] Begin INSERTing events as they come in — specifically, the GraphQL handler should add an extra call to insert the event at https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/user_usage_stats.go#L68:12, while leaving the existing storage in place as well (they will run in parallel for a short time).
- [ ] Update the GraphQL API call (defined at https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/schema.graphql#L268:5) to accept (1) arbitrary event names, (2) a `url` parameter, and (3) an `argument` parameter. 
- [ ] We currently heavily filter which frontend events get sent back through the GraphQL API in https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/web/src/tracking/services/serverAdminWrapper.tsx#L5:7. Going forward, both `trackPageView` and `trackAction` should call `logUserEvent(eventAction)` (instead of only doing so on certain types of events, and with modified event names). Pass the current URL through as the `url`, and nil/empty string as the `argument`
- [ ] Lastly, I'd recommend adding a feature flag to disable (or enable?) this functionality. 

And I think that's all I have. @nicksnyder any thoughts?

**Comment by nicksnyder** [comment]
LGTM. We should definitely protect this behind a feature flag. @dadlerj I think it is your call as to whether you want this new code path to be on for all customer by default or if you want to have it off by default and get certain customers to try it out.

**Comment by unknwon** [comment]
@dadlerj 

>We currently heavily filter which frontend events get sent back through the GraphQL API in https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/web/src/tracking/services/serverAdminWrapper.tsx#L5:7. Going forward, both `trackPageView` and `trackAction` should call `logUserEvent(eventAction)` (instead of only doing so on certain types of events, and with modified event names). Pass the current URL through as the `url`, and nil/empty string as the `argument`

Do you want this behavior to be added in this iteration (i.e. 3.8)?

>as well as any supporting service infrastructure, such as the gc to dump data after 93 days

Do you want hard or soft delete for the "expired" data after 93 days?

**Comment by nicksnyder** [comment]
> Do you want hard or soft delete for the "expired" data after 93 days?

I think hard delete is the only thing that makes sense. The only reason to delete is to save space.

**Comment by nicksnyder** [comment]
> Do you want this behavior to be added in this iteration (i.e. 3.8)?

Since it seems like we have time, let's try to get this in too in your current PR.

**Comment by unknwon** [comment]
Closed as implemented.
